crates/league-modpkg/src/read.rs

```diff
@@ -0,0 +1,89 @@
+use byteorder::{ReadBytesExt as _, LE};
+use io_ext::ReaderExt as _;
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    io::{BufReader, Read},
+};
+
+use crate::{error::ModpkgError, Modpkg, ModpkgAuthor, ModpkgChunk, ModpkgLicense};
+
+impl Modpkg {
+    pub const MAGIC: u64 = u64::from_le_bytes(*b"_modpkg_");
+
+    pub fn read(reader: &mut BufReader<impl Read>) -> Result<Self, ModpkgError> {
```

Comment:
I think there is missed opportunity to separate metadata here (like we did in last format).
Some kind of simple CBOR/msgpack that can be read all at once easily without rest of data.

Reply:
I don't want to put metadata (magic, version...) at the end because then I have to stream the whole file along the network to verify the magic. This would be an issue for Runeforge specifically.

As for putting metadata and path list as dummy chunk, I don't like the idea that much. I feel like it would just create extra overhead for those trying to process the files. I think we can put the metadata into a `msgpack` buffer at the start of the file instead. For path list we can just dump the data at the end of the file.

Reply:
Technically you wouldn't need to stream (range requests are a thing in http) but yeah i agree that moving stuff to end just moves complexity for some processing. 👍 

The reason i wanted to put metadata in regular chunk is because it actually makes processing easier. Additionally this means if we add signature to the file, the signature can now verify both metadata and data by virtue of verifying list of chunks (which in turn verifies data of chunks with a hash).

If this doesn't workout for you i think then just keeping all sizes in header first would be good enough?
```
u64 magic
u32 size_of_header // size of header allow skiping any future extensions for simple readers
u32 size_of_metdata
u32 count_of_chunks
u32 size_of_signature
u32 size_of_auxiliary_data
u32 size_of_string_table

... rest of header, 

[u8 metadata msgpack]  // simple readers can just skip to this with size_of_header
[ChunkDesc chunk descriptors]
[u8 signature] // <= signature over metadata + chunks contents
[u8 auxiliary data msgpack] // <= store wad list and features tags here?
[u8 strings table] // <= maybe simple length prefix & null terminated list here (must end with empty entry)
[u8 chunk data] 
```
lmk what you think

Reply:
I'm a bit hesitant personally to use `msgpack` for also storing the auxiliary data but I guess it doesn't really matter, we'll probably even get smaller size due to how it stores the data

Reply:
doesn't have to be msgpack
you could also probably make strings be part of auxiliary data?

Reply:
Actually we might want to keep this simple and just implement features and wads as indices into string table 
like this:
```rs
pub struct ModpkgChunk {
    path_hash: u64,
    offset: u64,
    compressed_size: u64,
    uncompressed_size: u64,
    compressed_checksum: u64,
    uncompressed_checksum: u64,
    path_string_index: u32, // unhashed path if any
    features_string_index: u32, // semicolon separated list of features this file belongs to
    wads_string_index: u32, // semicolon separated list of wad paths this file belongs to
    reserved1: u32, // necessary for alignment, make sure this is allways 0 in current version
}
```

Reply:
Yes that's how I wanted to do it as well